File: dawn_python/succinct.py
class CompactSuffixArray:
    def __init__(self, text):
        self.text = text + '$'
        self.rank_of_char = self.compute_rank_of_char()
        self.ranks = [self.rank_suffix(i) for i in range(len(self.text))]
        self.indices = sorted(range(len(self.text)), key=lambda i: self.ranks[i])

    def rank_suffix(self, i):
        n = len(self.text)
        r = 0
        for j in range(i, n):
            r = r * len(self.rank_of_char) + self.rank_of_char[ord(self.text[j])]
        return r

    def compute_rank_of_char(self):
        rank_of_char = [0] * 256
        used = [False] * 256
        rank = 0
        for c in self.text:
            if not used[ord(c)]:
                rank_of_char[ord(c)] = rank
                rank += 1
                used[ord(c)] = True
        return rank_of_char

    def compact_suffix_array(self):
        return self.indices


# Succinct static data structures, Jacobsen '89
class BitVector:
    k = 6
    
    def __init__(self, bits):
        self.bits = bits
        n = len(bits)
        self.rank_table = [0] * (n + 1)
        self.select_table = [-1] * (n + 1)

        for i in range(n + 1):
            if i % (1 << self.k) == 0:
                self.rank_table[i] = 0
                self.select_table[i] = -1
            else:
                prev = i - (i % (1 << self.k))
                self.rank_table[i] = self.rank_table[prev] + self.rank_naive(bits, i) - self.rank_naive(bits, prev)
                self.select_table[i] = self.select_table[prev] + self.select_naive(bits, i) - self.select_naive(bits, prev)

    def rank_naive(self, bits, i):
        return sum(1 for b in bits[:i + 1] if b)

    def select_naive(self, bits, i):
        cnt, j = 0, 0
        for j in range(len(bits)):
            if bits[j]:
                cnt += 1
                if cnt == i + 1:
                    return j
        return -1

    def rank(self, i):
        return self.rank_table[i]

    def select(self, i):
        l, r = 0, len(self.bits)
        while l <= r:
            m = (l + r) // 2
            if self.rank(m) < i:
                l = m + 1
            else:
                r = m - 1
        return l if l < len(self.bits) and self.rank(l) == i else -1

File: dawn_go/succinct.go
package compactsuffix

import "fmt"

func rankSuffix(text string, rankOfChar func(rune) int, n, i int) int {
	r := 0
	for j := i; j < n; j++ {
		r = r*len(rankOfChar) + rankOfChar(rune(text[j]))
	}
	return r
}

func computeRankOfChar(text string) func(rune) int {
	rankOfChar := make([]int, 256)
	used := make([]bool, 256)
	rank := 0
	for _, c := range text {
		if !used[c] {
			rankOfChar[c] = rank
			rank++
			used[c] = true
		}
	}
	return func(r rune) int { return rankOfChar[r] }
}

func CompactSuffixArray(text string) []int {
	n := len(text)
	rankOfChar := computeRankOfChar(text)
	ranks := make([]int, n)
	for i := 0; i < n; i++ {
		ranks[i] = rankSuffix(text, rankOfChar, n, i)
	}
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	sortByRanks(ranks, indices)
	return indices
}

func sortByRanks(ranks, indices []int) {
	sort.Slice(indices, func(i, j int) bool {
		return ranks[indices[i]] < ranks[indices[j]]
	})
}



// Succinct static data structures, Jacobsen '89
func rankNaive(bits []bool, i int) int {
	cnt := 0
	for j := 0; j <= i; j++ {
		if bits[j] {
			cnt++
		}
	}
	return cnt
}

func selectNaive(bits []bool, i int) int {
	cnt, j := 0, 0
	for ; cnt < i; j++ {
		if j >= len(bits) {
			return -1
		}
		if bits[j] {
			cnt++
		}
	}
	if cnt == i {
		return j - 1
	}
	return -1
}

type BitVector struct {
	bits        []bool
	rankTable   []int
	selectTable []int
	k int
}

func createBitVector(bits []bool) *BitVector {
	n := len(bits)
	rankTable := make([]int, n+1)
	selectTable := make([]int, n+1)

	for i := 0; i <= n; i++ {
		if i%(1<<k) == 0 {
			rankTable[i] = 0
			selectTable[i] = -1
		} else {
			rankTable[i] = rankTable[i-(i%(1<<k))] + rankNaive(bits, i) - rankNaive(bits, i-(i%(1<<k)))
			selectTable[i] = selectTable[i-(i%(1<<k))] + selectNaive(bits, i) - selectNaive(bits, i-(i%(1<<k)))
		}
	}

	return &BitVector{bits, rankTable, selectTable}
}

func (bv *BitVector) rank(i int) int {
	return bv.rankTable[i]
}

func (bv *BitVector) select(i int) int {
	l, r := 0, len(bv.bits)
	for l <= r {
		m := l + (r-l)/2
		if bv.rank(m) < i {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	return l
}